### PR TITLE
Add datepicker for scheduling gift cards

### PR DESCRIPTION
**Pull request by antiBaconMachine**
### PR Summary: 

Adds date picker for scheduling gift card sending for a future date.

### Why are these changes introduced?

Closes a great number of merchant frustrations

### What approach did you take?

We're using a native date picker for now as this was deemed to be the best starting point int terms of utility and accessibility. As the feature matures we may choose to implement something custom which is more themable but this is a good starting position.

Other theme developers may use a different date picker as long as the line itmem property is in ISO format YYYY-MM-DD

We also now use error messages which contain the field as part of the error string which is much better for localisation. Unfortunately it would be a breaking change to introduce this, so we've made it opt in my keying on the new _shopify_offset parameter. It's a bit strange to use this property as the switch but it works.

We want the TZ offset so that we can schedule the gift card send for a sensible time in the user's local timezone rather than in the shop timezone which might cause confusion.

Whilst I was here I also cleaned up a minor bug with empty line item properties being sent even when the checkbox was not checked. I've done this simply by disabling the fields whenever the checkbox is unchecked.

### Decision log

| # | Decision | Alternatives | Rationale | Downsides |
|---|---|---|---|---|
| 1 |   |   |   |   |

### Testing steps/scenarios

- [ ] Step 1

### Demo links

- [Store]( https://shop1.shopify.scheduling-theme-with-validation.ollie-edwards.eu.spin.dev/products/gift-card)
- [Editor](https://shop1.shopify.scheduling-theme-with-validation.ollie-edwards.eu.spin.dev/admin/themes/1/editor)

### Checklist
- [ ] Added PR summary for [release notes](https://themes.shopify.com/themes/dawn/styles/default#ReleaseNotes)
- [ ] Requested review from UX (Only for changes that are affecting the experience or perceivable visual details)
- [ ] Created a ticket for the [help.shopify.com](https://help.shopify.com) documentation team about updates to theme settings. (Internal-only task)
- [ ] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [ ] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [ ] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [ ] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [ ] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

### CODE REVIEW

File: assets/recipient-form.js
Comment (ludoboludo) [inline assets/recipient-form.js:71]: More nitpick for readability: 

```javascript
onChange() {
  if (this.checkboxInput.checked) {
    this.enableInputFields();
  } else {
    this.clearInputFields();
    this.disableInputFields();
    this.clearErrorMessage();
  }
}

```

File: assets/recipient-form.js
Comment (ludoboludo) [inline assets/recipient-form.js:14]: Out of curiosity, is there any back end validation/erroring if the date entered isn't in the future?

Say I make a mistake and enter `2022` as the year.

File: assets/recipient-form.js
Comment (ludoboludo) [inline assets/recipient-form.js:96]: Could this be something to add to `this.clearInputFields()` instead of its own function?

I see that `this.clearInputFields()` is run twice and in one instance, `this.disableInputFields()` isn't run, in the `resetRecipientForm()`. Would it make sense to also run it there ?

File: snippets/gift-card-recipient-form.liquid
Comment (ludoboludo) [inline snippets/gift-card-recipient-form.liquid:109]: When there is an error, the messaging is pretty vague.That seem to be because I'm testing locally. 
Shouldn't every store behave the same?
Here is what I'm getting:

**Review by ludoboludo — COMMENTED** [review]
A few comments, mostly nipticks. A few questions as well to make sure I understand the context 🙂 

Nothing blocking. 👍

File: snippets/gift-card-recipient-form.liquid
Comment (antiBaconMachine) [reply snippets/gift-card-recipient-form.liquid:231]: Yep it's as you say, the difference between UTC and local time. It exists because UX wanted the user's date input to be interpreted from their perspective rather than from the shops's perspective. It means the gift card is sent on the morning of the day selected in the user's timezone. If we didn't do it then we could only guarantee +- 1 day. In terms of naming, yes it's still changeable. It's an internal detail that users of flagship themes should basically never be aware of, but it could make a difference to other theme developers.

File: assets/recipient-form.js
Comment (antiBaconMachine) [reply assets/recipient-form.js:14]: Yeah, we only allow from today until 90 days currently, though that is an arbitrary choice.

File: assets/recipient-form.js
Comment (antiBaconMachine) [reply assets/recipient-form.js:96]: It's different because we never want to clear the offset field but we do want to disable it. It's because offset is not user input but it does sit along side user input and use the same line item property mechanism, so there are two similar use cases but not the same.

File: snippets/gift-card-recipient-form.liquid
Comment (antiBaconMachine) [reply snippets/gift-card-recipient-form.liquid:109]: So this is an temporary edge case that you should only see if javascript is disabled. That's clearly not the case in this screenshot so very interested in a replication for how you got this.

File: snippets/gift-card-recipient-form.liquid
Comment (ludoboludo) [reply snippets/gift-card-recipient-form.liquid:231]: Great sounds good. I think we could match the name of the JS method: `getTimezoneOffset`. So maybe `Recipient-timezone-offset-{{ section.id }}`

File: assets/recipient-form.js
Comment (ludoboludo) [reply assets/recipient-form.js:14]: Sounds good 👍

File: snippets/gift-card-recipient-form.liquid
Comment (antiBaconMachine) [reply snippets/gift-card-recipient-form.liquid:109]: Oh that's fine, the backend that supports the new error messages isn't merged yet. It should look better in the spin links above.

For some more context we originally just copied what the customer form does which is concatenate error messages with their field name in liquid like `{{ name }}{{ message }}` which we knew was bad but we were pressed for time so we didn't mess with the precedent. Now we have some time to revaluate we think it's pretty unacceptable to do this so we're fixing the messages so they are localised properly. That would be a breaking change if not handled properly so we're using the presence of the offset field as a hint the theme can cope with the new better error strings.

**Review by ludoboludo — APPROVED** [review]
I requested translations 👍 so we will have to wait on them before being able to merge

**Pull request merged**
